src/dso_api/dbroles.py

```diff
@@ -0,0 +1,143 @@
+            cls.activate_end_user(default_connection, log_source="set_end_user")
+
+    @classmethod
+    def activate_end_user(cls, user_connection: BaseDatabaseWrapper, *, log_source: str):
+        """Switch to the end-user role in the database."""
+        if not settings.DATABASE_SET_ROLE:
+            # Feature flag for previous hosting location
+            logger.debug("%s: End-user feature disabled (DATABASE_SET_ROLE=false)", log_source)
+            return
+
+        if getattr(user_connection, "_active_user_role", None):
+            logger.debug("%s: End-user already set, no need to switch roles again", log_source)
+            return
+
+        user_email = getattr(cls.current_user, "email", None)
+        if not user_email:
+            logger.debug("%s: No end-user email, not switching database roles", log_source)
+            cls._set_role(user_connection, ANONYMOUS_ROLE, ANONYMOUS_APP_NAME)
+            return
+
+        logger.debug("%s: Activating end-user context for %s", log_source, user_email)
+
+        # Log which role was activated
+        # Define which role to switch to
+        role_name = USER_ROLE.format(user_email=user_email)
+        try:
+            # BBN2: Exact account for specific access.
+            cls._set_role(user_connection, role_name, user_email)
+        except DataError as e:
+            # The role didn't exist.
+            if is_internal(user_email):
+                # BBN1: Internal employee, no specific account
+                cls._set_role(user_connection, INTERNAL_ROLE, user_email)
+            else:
+                logger.exception("External user %s has no database role %s", user_email, role_name)
+                raise PermissionError(f"User {user_email} is not available in database") from e
+
+    @classmethod
+    def _set_role(cls, user_connection: BaseDatabaseWrapper, role_name: str, app_name: str):
+        # By starting a transaction, any collection pooling (e.g. PgBouncer)
+        # can also ensure the connection is not reused for another session.
+        with user_connection.cursor() as c:
+            try:
+                c.execute(
+                    "BEGIN TRANSACTION; SET LOCAL ROLE %s; set application_name to %s;",
+                    (role_name, app_name),
+                )
+            except DatabaseError as e:
+                logger.debug("Switch role failed for %s: %s", role_name, e)
+                c.execute("ROLLBACK;")
+                raise
+            else:
+                setattr(user_connection, "_active_user_role", role_name or "NONE")
+                logger.info("Activated end-user database role '%s' for '%s'", role_name, app_name)
+
+    @classmethod
+    def deactivate_end_user(cls):
+        """Rollback the transaction when the local user was activated."""
+        user_email = getattr(cls.current_user, "email", None)
+        if not user_email or not setattr(default_connection, "_active_user_role", None):
```

Comment:
shouldnt this be `getattr`? otherwise it always evaluates to True

Reply:
yikes. indeed. I've wrapped all getattr() logic in separate functions now,